sled-agent/src/nexus.rs

```diff
@@ -202,3 +167,816 @@ impl ConvertInto<nexus_client::types::DatasetKind>
+        handle.notify_nexus_about_self(log).await;
+        let status = wait_for::<_, (), _, _>(
+            || async {
+                let status = handle.get_status().await.unwrap();
+                if !status.has_pending_notification {
+                    Ok(status)
+                } else {
+                    Err(CondCheckError::NotYet)
+                }
+            },
+            &Duration::from_millis(2),
+            &Duration::from_secs(15),
+        )
+        .await
+        .expect("Failed to get status from Nexus");
+
+        assert_eq!(status.total_get_requests_started, 1u64);
+        assert_eq!(status.total_get_requests_completed, 1u64);
+        assert_eq!(status.total_put_requests_started, 2u64);
+        assert_eq!(status.total_put_requests_completed, 2u64);
+        assert_eq!(status.has_outstanding_request, false);
+        assert_eq!(status.cancelled_pending_notifications, 3);
+        let expected = latest_sled_agent_info.lock().unwrap().clone();
+        assert_eq!(
+            status.nexus_known_info,
+            Some(NexusKnownInfo::Found(expected)),
+        );
+
+        // Inject a put error and trigger a put to nexus after updating the VMM
+        // reservoir size. It should eventually succeed.
+        put_error.store(true, Ordering::SeqCst);
+        {
+            let mut info = latest_sled_agent_info.lock().unwrap();
+            info.reservoir_size = (2 * 1024 * 1024u64).into();
+            info.generation = info.generation.next();
+        }
+        handle.notify_nexus_about_self(log).await;
+        // Wait for a steady state, when the the latest info has been put to nexus.
+        // Ensure the second get request has been sent.
+        let status = wait_for::<_, (), _, _>(
+            || async {
+                let status = handle.get_status().await.unwrap();
+                if !status.has_pending_notification
+                    && status.total_get_requests_started == 2
+                {
+                    Ok(status)
+                } else {
+                    Err(CondCheckError::NotYet)
+                }
+            },
+            &Duration::from_millis(2),
+            &Duration::from_secs(15),
+        )
+        .await
+        .expect("Failed to get status from Nexus");
+
+        // One extra get and put request because of the put error
+        assert_eq!(status.total_get_requests_completed, 2u64);
+
+        assert_eq!(status.total_put_requests_started, 4u64);
+        assert_eq!(status.total_put_requests_completed, 4u64);
+        assert_eq!(status.has_outstanding_request, false);
+        assert_eq!(status.cancelled_pending_notifications, 4);
+        let expected = latest_sled_agent_info.lock().unwrap().clone();
+        assert_eq!(
+            status.nexus_known_info,
+            Some(NexusKnownInfo::Found(expected)),
+        );
+
+        // Inject a get error and trigger a put to nexus after updating the VMM
```

Comment:
I did a double take on this comment and just want to check my understanding: is the idea here that because the PUT should be successful, there should be no GET, and therefore this error shouldn't impact how the task behaves? If that's right, does setting the error bear any load given that the number of get requests is checked below on 923-924?

EDIT: Ah, I see this becomes load-bearing even further below at line 937. Maybe worth a comment update here; it's up to you.

Reply:
Good idea. Comment changed in [4fd1afd](https://github.com/oxidecomputer/omicron/pull/5124/commits/4fd1afdd331707230daf942e5904f427122cac62)